core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/scheduling/WorkflowPipelinedRegionsBuilder.scala

```diff
@@ -333,42 +209,45 @@ class WorkflowPipelinedRegionsBuilder(
 
   private def populateTerminalOperatorsForBlockingLinks(): Unit = {
     val regionTerminalOperatorInOtherRegions =
-      new mutable.HashMap[PipelinedRegion, ArrayBuffer[OperatorIdentity]]()
-    getTopologicallyOrderedOperators().foreach(opId => {
-      val upstreamOps = getDirectUpstreamOperators(opId)
-      upstreamOps.foreach(upstreamOp => {
-        val linkFromUpstreamOp = LinkIdentity(
-          operatorToOpExecConfig(upstreamOp).topology.layers.last.id,
-          operatorToOpExecConfig(opId).topology.layers.head.id
-        )
-        if (operatorToOpExecConfig(opId).isInputBlocking(linkFromUpstreamOp)) {
-          val prevInOrderRegions = getPipelinedRegionsFromOperatorId(upstreamOp)
-          for (prevInOrderRegion <- prevInOrderRegions) {
-            if (
-              !regionTerminalOperatorInOtherRegions.contains(
-                prevInOrderRegion
-              ) || !regionTerminalOperatorInOtherRegions(prevInOrderRegion).contains(opId)
-            ) {
-              val terminalOps = regionTerminalOperatorInOtherRegions.getOrElseUpdate(
-                prevInOrderRegion,
-                new ArrayBuffer[OperatorIdentity]()
-              )
-              terminalOps.append(opId)
-              regionTerminalOperatorInOtherRegions(prevInOrderRegion) = terminalOps
+      new mutable.HashMap[PipelinedRegion, ArrayBuffer[LayerIdentity]]()
+    this.physicalPlan
+      .topologicalIterator()
+      .foreach(opId => {
+        val upstreamOps = this.physicalPlan.getUpstream(opId)
+        upstreamOps.foreach(upstreamOp => {
+          val linkFromUpstreamOp = LinkIdentity(
+            physicalPlan.operatorMap(upstreamOp).id,
+            physicalPlan.operatorMap(opId).id
+          )
+          if (physicalPlan.operatorMap(opId).isInputBlocking(linkFromUpstreamOp)) {
+            val prevInOrderRegions = getPipelinedRegionsFromOperatorId(upstreamOp)
+            for (prevInOrderRegion <- prevInOrderRegions) {
+              if (
+                !regionTerminalOperatorInOtherRegions.contains(
+                  prevInOrderRegion
+                ) || !regionTerminalOperatorInOtherRegions(prevInOrderRegion).contains(opId)
+              ) {
+                val terminalOps = regionTerminalOperatorInOtherRegions.getOrElseUpdate(
+                  prevInOrderRegion,
+                  new ArrayBuffer[LayerIdentity]()
+                )
+                terminalOps.append(opId)
+                regionTerminalOperatorInOtherRegions(prevInOrderRegion) = terminalOps
+              }
             }
           }
-        }
+        })
       })
-    })
 
     for ((region, terminalOps) <- regionTerminalOperatorInOtherRegions) {
-      region.blockingDowstreamOperatorsInOtherRegions = terminalOps.toArray
+      val newRegion = region.copy(blockingDownstreamOperatorsInOtherRegions = terminalOps.toArray)
+      replaceVertex(pipelinedRegionsDAG, region, newRegion)
```

Comment:
A clarification question: why couldn't the old code, `region.blockingDowstreamOperatorsInOtherRegions = terminalOps.toArray`, be preserved? Did it not actually update the `pipelinedRegionsDAG` since `region` is passed as a copy? 
Also Is there a specific reason you have to write `replaceVertex` for this? Is `DiretedAcyclicGraph`'s vertices not modifiable?

Reply:
1. `PipelinedRegion` is changed to a case class, which means all members written in the constructor are immutable class members and cannot be reassigned. so  `region.blockingDowstreamOperatorsInOtherRegions = terminalOps.toArray` can not compile. This is part of the effort to promote immutability in the code.
2. yes the library doesn't have an API to replace a vertex, so I have to construct a new dag to do the replacement, see https://stackoverflow.com/questions/8766741/changing-contents-of-vertex-with-jgrapht